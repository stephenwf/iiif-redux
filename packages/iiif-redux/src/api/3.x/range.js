import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationCollections,
  getAllAnnotationPages,
  getAllAnnotations,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllSelectors,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const range = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getBehavior
   * - getViewingDirection
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );
  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id] || null;
    }
  );

  const getSupplementaryId = createSelector(selector, linking.getSupplementary);
  const getSupplementary = createSelector(
    getSupplementaryId,
    getAllAnnotationCollections,
    mapById
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getRanges
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(
    getItemIds,
    getAllAnnotations,
    getAllCanvases,
    getAllRanges,
    getAllResources,
    getAllCanvasReferences,
    getAllSelectors,
    (
      itemIds,
      allAnnotations,
      allCanvases,
      allRanges,
      allResources,
      allCanvasReferences,
      allSelectors
    ) => {
      const map = {
        annotation: allAnnotations,
        canvas: allCanvases,
        range: allRanges,
        resource: allResources,
        canvasReference: allCanvasReferences,
        selector: allSelectors,
      };
      return itemIds.map(({ schema, id }) => map[schema][id]);
    }
  );

  const getRangeIds = createSelector(getItemIds, rangeIds =>
    rangeIds
      .filter(({ schema }) => schema === 'range')
      .map(singleRange => singleRange.id)
  );
  const getRanges = createSelector(getRangeIds, getAllRanges, mapAllById);

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    // Technical.
    getId,
    getType,
    getBehavior,
    getViewingDirection,

    // Descriptive.
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,

    // Linking
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getSupplementaryId,
    getSupplementary,

    // Structural.
    getItemIds,
    getItems,
    getRangeIds,
    getRanges,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(range);

export default range;
