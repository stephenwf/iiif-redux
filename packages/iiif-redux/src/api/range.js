import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import { getAllRanges } from './all';
import presentation2Range from './2.x/range';

const range = presentation2Range;

export default range;

export const ranges = resourceListSelectorFactory(getAllRanges, range);

export const rangeByIdSelector = byIdSelectorFactory(range, 'ranges');
