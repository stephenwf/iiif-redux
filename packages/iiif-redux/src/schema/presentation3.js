import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';
import parseSelectorTarget from '../utility/parseTargetSelector';

function createEntity(name, processStrategy) {
  const options = {
    idAttribute: 'id',
  };
  if (processStrategy) {
    options.processStrategy = processStrategy;
  }

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations', original => {
  const body = Array.isArray(original.body) ? original.body : [original.body];
  const target = Array.isArray(original.target)
    ? original.target
    : [original.target];
  const newObj = { ...original };
  if (original.body) {
    newObj.body = body;
  }
  if (original.target) {
    newObj.target = target;
  }
  return newObj;
}); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollections'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReferences'); // 10
const selector = new schema.Entity(
  'selectors',
  {
    id: canvas,
  },
  {
    processStrategy(value, parent, key) {
      // @todo This is an impossible path to hit. The process strategy is a strange function that
      // doesn't always hit the containing code. Will need to review at a later date.
      // if (typeof value === 'string') {
      //   return parseSelectorTarget(value);
      // }
      const id = value['@id'] || value.id;
      if (id) {
        return parseSelectorTarget(id);
      }
      return value;
    },
  }
);

// Unofficial types.
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    contentResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'contentResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    selector,
    canvasReference,
    canvas,
  },
  entity => {
    if (entity && entity.id && entity.id.indexOf('#') !== -1) {
      return 'selector';
    }
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBodyMappings = {
  Choice: 'choice',
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const annotationBody = new schema.Union(
  {
    contentResource,
    choice,
  },
  input => {
    if (annotationBodyMappings[input.type]) {
      return annotationBodyMappings[input.type];
    }
    return 'contentResource';
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    selector,
    range,
    canvasReference,
  },
  input => {
    if (input && input.id && input.id.indexOf('#') !== -1) {
      return 'selector';
    }

    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: [annotationBody],
  target: [selector],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  rendering: [contentResource],
  partOf: partOf,
  // Structural
  annotations: [annotationPage],
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
