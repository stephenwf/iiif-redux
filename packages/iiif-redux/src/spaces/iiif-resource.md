# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
