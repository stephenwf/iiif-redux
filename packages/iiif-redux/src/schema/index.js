import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
