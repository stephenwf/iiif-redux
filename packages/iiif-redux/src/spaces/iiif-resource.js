import cachedFetch from 'fetch-unless-cached';
import { createActions, handleActions } from 'redux-actions';
import validUrl from 'valid-url';
import { call, put, select, takeEvery, all } from 'redux-saga/effects';
import deepmerge from 'deepmerge';
import update from 'immutability-helper';
import * as presentation2 from '../schema/presentation2';
import * as presentation3 from '../schema/presentation3';
import trackPresentationVersion from '../utility/trackPresentationVersion';
import {
  COLLECTION_ERROR,
  COLLECTION_REQUEST,
  COLLECTION_SUCCESS,
} from './collections';
import {
  MANIFEST_ERROR,
  MANIFEST_REQUEST,
  MANIFEST_SUCCESS,
} from './manifests';
import getPresentationVersionFromResource from '../utility/getPresentationVersionFromResource';
const debug = require('debug')('iiif-redux');

const IIIF_RESOURCE_REQUEST = 'IIIF_RESOURCE_REQUEST';
const IIIF_RESOURCE_SUCCESS = 'IIIF_RESOURCE_SUCCESS';
const IIIF_RESOURCE_ERROR = 'IIIF_RESOURCE_ERROR';
const IIIF_RESOURCE_REQUEST_UNKNOWN = 'IIIF_RESOURCE_REQUEST_UNKNOWN';
const IIIF_RESOURCE_IDENTIFY = 'IIIF_RESOURCE_IDENTIFY';

const DEFAULT_MAPPINGS = {
  collection: [COLLECTION_REQUEST, COLLECTION_SUCCESS, COLLECTION_ERROR],
  manifest: [MANIFEST_REQUEST, MANIFEST_SUCCESS, MANIFEST_ERROR],
};

const {
  iiifResourceRequestUnknown,
  iiifResourceRequest,
  iiifResourceSuccess,
  iiifResourceError,
} = createActions({
  [IIIF_RESOURCE_REQUEST_UNKNOWN]: (
    resourceId,
    options = {},
    mappings = DEFAULT_MAPPINGS
  ) => ({
    resourceId,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
    mappings,
  }),
  [IIIF_RESOURCE_REQUEST]: (resourceId, types, schema, options = {}) => ({
    resourceId,
    types,
    schema,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
  }),
  [IIIF_RESOURCE_SUCCESS]: (resourceId, normalizedResponse) => ({
    resourceId,
    normalizedResponse,
  }),
  [IIIF_RESOURCE_ERROR]: (resourceId, error) => ({
    resourceId,
    error,
  }),
});

const DEFAULT_STATE = {
  dereferenced: {},
  resources: {
    collections: {},
    sequences: {},
    manifests: {},
    canvases: {},
    canvasReferences: {},
    annotations: {},
    annotationLists: {},
    annotationPages: {},
    annotationCollections: {},
    ranges: {},
    layers: {},
    imageResources: {},
    externalResources: {},
    contentResources: {},
    choices: {},
  },
};

const resourceReducer = handleActions(
  {
    [iiifResourceSuccess]: (state, { payload: { normalizedResponse } }) =>
      deepmerge(state, normalizedResponse),
  },
  DEFAULT_STATE.resources
);

const dereferencedReducer = handleActions(
  {
    [iiifResourceRequest]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            requested: state[resourceId]
              ? state[resourceId].requested
              : new Date(),
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),
    [iiifResourceRequestUnknown]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),

    [iiifResourceError]: (state, { payload: { resourceId, error } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
          error: { $set: error },
        },
      }),

    [iiifResourceSuccess]: (state, { payload: { resourceId } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
        },
      }),
  },
  DEFAULT_STATE.dereferenced
);

async function requestResource(resourceId, options) {
  /* istanbul ignore if */
  if (process.env.NODE_ENV === 'production') {
    return cachedFetch(resourceId);
  }

  return fetch(resourceId).then(resp => resp.json());
}

function* errorAction(type, resourceId, error) {
  yield put(iiifResourceError(resourceId, error));
  if (type) {
    yield put({ type, payload: { resourceId, error } });
  }
}

function* successAction(type, resourceId, normalizedResponse) {
  yield put(iiifResourceSuccess(resourceId, normalizedResponse));
  yield put({ type, payload: { resourceId, normalizedResponse } });
}

function* requestIiifResource({ payload }) {
  const { resourceId, types, ...options } = payload;
  if (types.length !== 3 && process.env.NODE_ENV !== 'production') {
    yield call(
      errorAction,
      null,
      resourceId,
      'DEV ERROR: You must pass in exactly 3 action types'
    );
  }

  const [REQUEST, SUCCESS, ERROR] = types;

  if (!validUrl.isWebUri(resourceId)) {
    yield call(
      errorAction,
      ERROR,
      resourceId,
      'Resource is not a valid Web URI.'
    );
    return;
  }

  const state = yield select();

  // @todo logic for cache.
  if (
    state.dereferenced[resourceId] &&
    state.dereferenced[resourceId].loading ===
      false /*&& options.forceFresh === false*/
  ) {
    debug('Skipping fetch for resource %s using cache.', resourceId);

    return;
  }

  yield put({ type: REQUEST, payload });

  try {
    debug('Fetching resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);

    const { result, entities } = yield call(
      importResource,
      response,
      resourceId
    );
    debug('Finished normalize resource %s', result.id);

    yield call(successAction, SUCCESS, result.id, entities);
  } catch (err) {
    debug('Error: %O', err);
    yield call(errorAction, ERROR, resourceId, err);
  }
}

function* importResource(response, resourceId) {
  const version = getPresentationVersionFromResource(response);
  if (version === 2) {
    return yield call(importPresentation2Resource, response, resourceId);
  }
  if (version === 3) {
    return yield call(importPresentation3Resource, response, resourceId);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* importPresentation2Resource(response, resourceId) {
  if (response['@id'] !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 2 resource, patching... Found: %s Expected: %s',
      response['@id'],
      resourceId
    );
    response['@id'] = resourceId;
  }

  debug('Starting normalize presentation 2 resource %s', resourceId);

  return trackPresentationVersion(
    2,
    presentation2.normalize(response, presentation2.resource)
  );
}

function* importPresentation3Resource(response, resourceId) {
  if (response.id !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 3 resource, patching... Found: %s Expected: %s',
      response.id,
      resourceId
    );
    response.id = resourceId;
  }

  debug('Starting normalize presentation 3 resource %s', resourceId);

  return trackPresentationVersion(
    3,
    presentation3.normalize(response, presentation3.resource)
  );
}

function* getResourceType(resourceId, response) {
  const presentationVersion = getPresentationVersionFromResource(response);
  if (presentationVersion === 2) {
    return yield call(getPresentation2ResourceType, resourceId, response);
  }
  if (presentationVersion === 3) {
    return yield call(getPresentation3ResourceType, resourceId, response);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* getPresentation2ResourceType(resourceId, response) {
  if (!presentation2.RESOURCE_TYPE_MAP[response['@type']]) {
    throw new Error('Unknown Presentation 2 resource type');
  }

  return presentation2.RESOURCE_TYPE_MAP[response['@type']];
}

function* getPresentation3ResourceType(resourceId, response) {
  if (!presentation3.RESOURCE_TYPE_MAP[response.type]) {
    throw new Error('Unknown Presentation 3 resource type');
  }

  return presentation3.RESOURCE_TYPE_MAP[response.type];
}

function* requestUnknownResource({
  payload: { resourceId, mappings, ...options },
}) {
  if (!validUrl.isWebUri(resourceId)) {
    yield put(iiifResourceError(resourceId, 'Resource is not a valid URL.'));
    return;
  }

  try {
    // 1) Request the resource.
    debug('Fetching unknown resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);
    if (!response) {
      yield put(iiifResourceError(resourceId, 'Unknown resource'));
      return;
    }

    const type = yield call(getResourceType, resourceId, response);

    if (!mappings[type]) {
      yield put(
        iiifResourceError(
          resourceId,
          `Resource type is not in configured mappings (${Object.keys(
            mappings
          ).join(', ')})`
        )
      );
      return;
    }

    // 3) Grab constants from a map (to be created).
    const [REQUEST, SUCCESS, ERROR] = mappings[type];

    // 4) Send to regular import (will be cached).
    yield put(iiifResourceRequest(resourceId, [REQUEST, SUCCESS, ERROR]));
  } catch (err) {
    yield put(iiifResourceError(resourceId, err.message));
  }
}

function* saga() {
  yield all([
    takeEvery(IIIF_RESOURCE_REQUEST, requestIiifResource),
    takeEvery(IIIF_RESOURCE_REQUEST_UNKNOWN, requestUnknownResource),
  ]);
}

export {
  saga,
  resourceReducer,
  dereferencedReducer,
  iiifResourceRequest,
  iiifResourceRequestUnknown,
  DEFAULT_STATE,
  IIIF_RESOURCE_REQUEST,
  IIIF_RESOURCE_SUCCESS,
  IIIF_RESOURCE_ERROR,
  IIIF_RESOURCE_REQUEST_UNKNOWN,
  IIIF_RESOURCE_IDENTIFY,
};
