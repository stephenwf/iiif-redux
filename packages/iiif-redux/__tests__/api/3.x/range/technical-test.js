import range from '../../../../src/api/3.x/range';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/range/technical', () => {
  describe.each(fixtures)('Fixture: %s', (name, fixture) => {
    const {
      result: { id },
      entities,
    } = normalize(fixture);
    const state = {
      resources: { ...DEFAULT_STATE.resources, ...entities },
    };
    const selectRange = s => s.resources.ranges[id];

    const { getId, getType, getBehavior, getViewingDirection } = range(
      selectRange
    );

    test('getId', () => {
      expect(getId(state)).toMatchSnapshot(`getId ${name}`);
    });

    test('getType', () => {
      expect(getType(state)).toMatchSnapshot(`getType ${name}`);
    });

    test('getBehavior', () => {
      expect(getBehavior(state)).toMatchSnapshot(`getBehavior ${name}`);
    });

    test('getViewingDirection', () => {
      expect(getViewingDirection(state)).toMatchSnapshot(
        `getViewingDirection ${name}`
      );
    });
  });
});
