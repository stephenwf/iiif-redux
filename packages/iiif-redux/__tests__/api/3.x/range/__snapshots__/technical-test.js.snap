// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/range/technical Fixture: Full range getBehavior: getBehavior Full range 1`] = `
Array [
  "hidden",
]
`;

exports[`api/3.x/range/technical Fixture: Full range getId: getId Full range 1`] = `"https://example.org/iiif/range/1"`;

exports[`api/3.x/range/technical Fixture: Full range getType: getType Full range 1`] = `"Range"`;

exports[`api/3.x/range/technical Fixture: Full range getViewingDirection: getViewingDirection Full range 1`] = `"right-to-left"`;

exports[`api/3.x/range/technical Fixture: Range from IIIF spec getBehavior: getBehavior Range from IIIF spec 1`] = `null`;

exports[`api/3.x/range/technical Fixture: Range from IIIF spec getId: getId Range from IIIF spec 1`] = `"https://example.org/iiif/book1/range/r0"`;

exports[`api/3.x/range/technical Fixture: Range from IIIF spec getType: getType Range from IIIF spec 1`] = `"Range"`;

exports[`api/3.x/range/technical Fixture: Range from IIIF spec getViewingDirection: getViewingDirection Range from IIIF spec 1`] = `"left-to-right"`;

exports[`api/3.x/range/technical Fixture: Range minimal getBehavior: getBehavior Range minimal 1`] = `null`;

exports[`api/3.x/range/technical Fixture: Range minimal getId: getId Range minimal 1`] = `"https://example.org/iiif/range/1"`;

exports[`api/3.x/range/technical Fixture: Range minimal getType: getType Range minimal 1`] = `"Range"`;

exports[`api/3.x/range/technical Fixture: Range minimal getViewingDirection: getViewingDirection Range minimal 1`] = `"left-to-right"`;

exports[`api/3.x/range/technical Fixture: Range with start property getBehavior: getBehavior Range with start property 1`] = `null`;

exports[`api/3.x/range/technical Fixture: Range with start property getId: getId Range with start property 1`] = `"https://example.org/iiif/range/1"`;

exports[`api/3.x/range/technical Fixture: Range with start property getType: getType Range with start property 1`] = `"Range"`;

exports[`api/3.x/range/technical Fixture: Range with start property getViewingDirection: getViewingDirection Range with start property 1`] = `"left-to-right"`;
